Spigot-Server-Patches/0299-Slime-Patherfinder-Events.patch

```diff
@@ -0,0 +1,123 @@
+-                this.b = (float) this.a.getRandom().nextInt(360);
++                // Paper start
++                com.destroystokyo.paper.event.entity.SlimeChangeDirectionEvent event = new com.destroystokyo.paper.event.entity.SlimeChangeDirectionEvent((org.bukkit.entity.Slime) this.a.getBukkitEntity(), (float) this.a.getRandom().nextInt(360));
++                if (!this.a.canWander || !event.callEvent()) return;
++                this.b = event.getNewYaw();
++                // Paper end
+             }
+ 
+             ((EntitySlime.ControllerMoveSlime) this.a.getControllerMove()).a(this.b, false);
+@@ -377,7 +381,16 @@ public class EntitySlime extends EntityInsentient implements IMonster {
+         public boolean a() {
+             EntityLiving entityliving = this.a.getGoalTarget();
+ 
+-            return entityliving == null ? false : (!entityliving.isAlive() ? false : !(entityliving instanceof EntityHuman) || !((EntityHuman) entityliving).abilities.isInvulnerable);
++            // Paper start
++            if (entityliving != null && entityliving.isAlive() && (!(entityliving instanceof EntityHuman) || !((EntityHuman) entityliving).abilities.isInvulnerable)) {
++                if (this.a.canWander && new com.destroystokyo.paper.event.entity.SlimeTargetLivingEntityEvent((org.bukkit.entity.Slime) this.a.getBukkitEntity(), (org.bukkit.entity.LivingEntity) entityliving.getBukkitEntity()).callEvent()) {
++                    return true;
++                }
++                this.b = 0;
++                this.a.setGoalTarget(null);
++            }
++            return false;
++            // Paper end
+         }
+ 
+         public void c() {
+@@ -388,7 +401,16 @@ public class EntitySlime extends EntityInsentient implements IMonster {
+         public boolean b() {
+             EntityLiving entityliving = this.a.getGoalTarget();
+ 
+-            return entityliving == null ? false : (!entityliving.isAlive() ? false : (entityliving instanceof EntityHuman && ((EntityHuman) entityliving).abilities.isInvulnerable ? false : --this.b > 0));
++            // Paper start
++            if (entityliving != null && entityliving.isAlive() && (!(entityliving instanceof EntityHuman) || !((EntityHuman) entityliving).abilities.isInvulnerable)) {
++                if (this.a.canWander && new com.destroystokyo.paper.event.entity.SlimeTargetLivingEntityEvent((org.bukkit.entity.Slime) this.a.getBukkitEntity(), (org.bukkit.entity.LivingEntity) entityliving.getBukkitEntity()).callEvent()) {
++                    return true;
++                }
++                this.b = 0;
++                this.a.setGoalTarget(null);
++            }
++            return false;
++            // Paper end
+         }
+ 
+         public void e() {
+@@ -452,4 +474,16 @@ public class EntitySlime extends EntityInsentient implements IMonster {
+             }
+         }
+     }
++
++    // Paper start
++    private boolean canWander = true;
```

Comment:
We should persist this to a Paper.CanWander NBT Tag

Reply:
@billygalbreath able to make the change?

Reply:
Done ^_^